src/core/thread/osthread.d

```diff
@@ -2608,121 +2617,71 @@ else
-                __asm("std  17, $0", "=*m", regs.ptr +  3);
-                __asm("std  18, $0", "=*m", regs.ptr +  4);
-                __asm("std  19, $0", "=*m", regs.ptr +  5);
-                __asm("std  20, $0", "=*m", regs.ptr +  6);
-                // Work around LLVM bug 21443 (http://llvm.org/bugs/show_bug.cgi?id=21443)
-                // Because we clobber r0 a different register is choosen
-                __asm("std  21, $0", "=*m,~{r0}", regs.ptr +  7);
-                __asm("std  22, $0", "=*m", regs.ptr +  8);
-                __asm("std  23, $0", "=*m", regs.ptr +  9);
-                __asm("std  24, $0", "=*m", regs.ptr + 10);
-                __asm("std  25, $0", "=*m", regs.ptr + 11);
-                __asm("std  26, $0", "=*m", regs.ptr + 12);
-                __asm("std  27, $0", "=*m", regs.ptr + 13);
-                __asm("std  28, $0", "=*m", regs.ptr + 14);
-                __asm("std  29, $0", "=*m", regs.ptr + 15);
-                __asm("std  30, $0", "=*m", regs.ptr + 16);
-                __asm("std  31, $0", "=*m", regs.ptr + 17);
-
-                __asm("std   1, $0", "=*m", &sp);
+                static foreach (i; 0 .. regs.length)
+                {{
+                    enum int j = 14 + i; // source register
+                    static if (j == 21)
+                    {
+                        // Work around LLVM bug 21443 (http://llvm.org/bugs/show_bug.cgi?id=21443)
+                        // Because we clobber r0 a different register is chosen
+                        asm { ("std "~j.stringof~", %0") : "=m" regs[i] : : "r0"; }
+                    }
+                    else
+                        asm { ("std "~j.stringof~", %0") : "=m" regs[i]; }
+                }}
+
+                asm { "std 1, %0" : "=m" sp; }
             }
             else version (AArch64)
             {
                 // Callee-save registers, x19-x28 according to AAPCS64, section
                 // 5.1.1.  Include x29 fp because it optionally can be a callee
                 // saved reg
                 size_t[11] regs = void;
-                __asm("stp x19, x20, $0", "=*m", regs.ptr + 0);
-                __asm("stp x21, x22, $0", "=*m", regs.ptr + 2);
-                __asm("stp x23, x24, $0", "=*m", regs.ptr + 4);
-                __asm("stp x25, x26, $0", "=*m", regs.ptr + 6);
-                __asm("stp x27, x28, $0", "=*m", regs.ptr + 8);
-                __asm("str x29, $0", "=*m", regs.ptr + 10);
-                sp = __asm!(void*)("mov $0, sp", "=r");
+                // store the registers in pairs
+                asm { "stp x19, x20, %0" : "=m" regs[ 0], "=m" regs[1]; }
+                asm { "stp x21, x22, %0" : "=m" regs[ 2], "=m" regs[3]; }
+                asm { "stp x23, x24, %0" : "=m" regs[ 4], "=m" regs[5]; }
+                asm { "stp x25, x26, %0" : "=m" regs[ 6], "=m" regs[7]; }
+                asm { "stp x27, x28, %0" : "=m" regs[ 8], "=m" regs[9]; }
```

Comment:
The 2nd output operator is unused in the template, but I've added it to let LLVM know that 2 registers are stored per asm statement.